pow/src/main/java/tech/pegasys/teku/pow/fallback/FallbackAwareEth1Provider.java

```diff
@@ -0,0 +1,97 @@
+import java.util.function.Function;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.web3j.protocol.core.methods.response.EthBlock.Block;
+import org.web3j.protocol.core.methods.response.EthCall;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.pow.Eth1Provider;
+import tech.pegasys.teku.pow.fallback.strategy.Eth1ProviderSelector;
+
+public class FallbackAwareEth1Provider implements Eth1Provider {
+  private static final Logger LOG = LogManager.getLogger();
+  private final Eth1ProviderSelector eth1ProviderSelector;
+
+  public FallbackAwareEth1Provider(final Eth1ProviderSelector eth1ProviderSelector) {
+    this.eth1ProviderSelector = eth1ProviderSelector;
+  }
+
+  @Override
+  public SafeFuture<Optional<Block>> getEth1Block(final UInt64 blockNumber) {
+    return run(eth1Provider -> eth1Provider.getEth1Block(blockNumber));
+  }
+
+  @Override
+  public SafeFuture<Optional<Block>> getEth1BlockWithRetry(
+      final UInt64 blockNumber, final Duration retryDelay, final int maxRetries) {
+    return run(
+        eth1Provider -> eth1Provider.getEth1BlockWithRetry(blockNumber, retryDelay, maxRetries));
+  }
+
+  @Override
+  public SafeFuture<Block> getGuaranteedEth1Block(final String blockHash) {
+    return run(eth1Provider -> eth1Provider.getGuaranteedEth1Block(blockHash));
+  }
+
+  @Override
+  public SafeFuture<Block> getGuaranteedEth1Block(final UInt64 blockNumber) {
+    return run(eth1Provider -> eth1Provider.getGuaranteedEth1Block(blockNumber));
+  }
+
+  @Override
+  public SafeFuture<Optional<Block>> getEth1Block(final String blockHash) {
+    return run(eth1Provider -> eth1Provider.getEth1Block(blockHash));
+  }
+
+  @Override
+  public SafeFuture<Optional<Block>> getEth1BlockWithRetry(
+      final String blockHash, final Duration retryDelay, final int maxRetries) {
+    return run(
+        eth1Provider -> eth1Provider.getEth1BlockWithRetry(blockHash, retryDelay, maxRetries));
+  }
+
+  @Override
+  public SafeFuture<Block> getLatestEth1Block() {
+    return run(Eth1Provider::getLatestEth1Block);
+  }
+
+  @Override
+  public SafeFuture<EthCall> ethCall(
+      final String from, final String to, final String data, final UInt64 blockNumber) {
+    return run(eth1Provider -> eth1Provider.ethCall(from, to, data, blockNumber));
+  }
+
+  @Override
+  public SafeFuture<BigInteger> getChainId() {
+    return run(Eth1Provider::getChainId);
+  }
+
+  private <T> SafeFuture<T> run(final Function<Eth1Provider, SafeFuture<T>> task) {
+    final SafeFuture<T> result = task.apply(eth1ProviderSelector.bestCandidate());
+    return result.catchAndRethrow(
+        throwable -> {
+          LOG.warn(
+              "Error caught while calling Eth1Provider, switching provider for next calls",
+              throwable);
+          eth1ProviderSelector.updateBestCandidate();
+        });
```

Comment:
I think this should be retrying the call on the next provider.  If you have a local node and infura configured, this will cause the first failure in your local node to trigger a fail over to infer and since infura won't go down it will never switch back.

I'd also actually do the retry with the next "ready" provider inside this class - that way the caller only gets an error if all providers are unavailable.

It would probably make sense to mark a provider as offline/unhealthy for some period if we get a failure from it to avoid repeatedly retrying calls to it before moving on to the next provider.  Might be a followup PR as it would work in really nicely with a periodic in-sync check - you'd mark a provider as offline if a request failed and mark it online again when the in-sync check next passes.

Reply:
Done. The retry is done directly inside this class.